src/Servers/Kestrel/Core/src/Internal/Infrastructure/HttpUtilities.cs

```diff
@@ -83,54 +84,63 @@ private static ulong GetMaskAsLong(ReadOnlySpan<byte> bytes)
 
     // The same as GetAsciiStringNonNullCharacters but throws BadRequest
     [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    public static unsafe string GetHeaderName(this ReadOnlySpan<byte> span)
+    public static string GetHeaderName(this ReadOnlySpan<byte> span)
     {
         if (span.IsEmpty)
         {
             return string.Empty;
         }
 
-        fixed (byte* source = &MemoryMarshal.GetReference(span))
+        var str = string.Create(span.Length, span, (destination, spanPtr) =>
         {
-            return string.Create(span.Length, new IntPtr(source), s_getHeaderName);
-        }
-    }
-
-    private static unsafe void GetHeaderName(Span<char> buffer, IntPtr state)
-    {
-        fixed (char* output = &MemoryMarshal.GetReference(buffer))
-        {
-            // This version of AsciiUtilities returns null if there are any null (0 byte) characters
-            // in the string
-            if (!StringUtilities.TryGetAsciiString((byte*)state.ToPointer(), output, buffer.Length))
+            if (Ascii.ToUtf16(spanPtr, destination, out _) != OperationStatus.Done)
             {
                 KestrelBadHttpRequestException.Throw(RequestRejectionReason.InvalidCharactersInHeaderName);
             }
+        });
+
+        if (str.Contains('\0'))
```

Comment:
Can this test be combined with the above one?
The exception thrown is the same, so it should make the method smaller (in machine code).